### PR TITLE
zlib-ng: remove overly specific -Wl,--undefined-version flag

### PR DESCRIPTION
**Pull request by haampie**
lld 17 and higher error by default when a symbol is versioned with a
non-existing version.

That change exposed various issues in different packages:

https://github.com/gentoo/gentoo/pull/29097#issuecomment-1383304538
https://github.com/termux/termux-packages/issues/18810
https://github.com/zlib-ng/zlib-ng/pull/1421

The issue was fixed for zlib-ng in 2.1.2, so adding the flag
`-Wl,--undefined-version` unconditionally only hides possible
regressions in the package going forward. Further, use of `%cce@17` may
not imply use of `lld` as a linker.

I would suggest not to deal with `-Wl,--(no-)undefined-version` flags in
Spack, because we don't control the choice of linker anyways -- it's not
in the dag & we don't add -fuse-ld / -ld-path / -B type of LDFLAGS.

Instead, fix the relevant packages if they fail to link.

### CODE REVIEW

**Review by alalazo — DISMISSED** [review]

**Comment by v-dobrev** [comment]
> Instead, fix the relevant packages if they fail to link.

I'm not sure what you mean here. The logic you are removing in this PR is fixing the package that fails to build correctly (`zlib-ng`) with the `cce@17` version (this is the default CCE version on the rzVernal machine at LLNL which has nodes similar to Frontier) that I was testing with in #44010 (see the relevant description of the issue here: https://github.com/spack/spack/pull/44010#discussion_r1628365884).

Also, in my tests, Spack was using `zlib-ng@2.1.6`, so the issue that I saw cannot be the one fixed v2.1.2 (unless the issue was re-introduced in v2.1.6 after it was fixed in 2.1.2).

If this logic is removed, I can work around the issue by adding the flag `-Wl,--undefined-version` to `ldflags` in my `compilers.yaml`, however every other user of Spack with `cce@17` will probably run into the same issue and will have to spend time figuring out the same or similar workaround. Therefore, I thought having this logic in Spack will be helpful.

File: var/spack/repos/builtin/packages/zlib-ng/package.py [hunk 1]
```diff
@@ -76,8 +76,6 @@ def libs(self):
     def flag_handler(self, name, flags):
         if name == "cflags" and self.spec.satisfies("+pic build_system=autotools"):
             flags.append(self.compiler.cc_pic_flag)
-        if name == "ldflags" and self.spec.satisfies("%cce@17"):
-            flags.append("-Wl,--undefined-version")
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/zlib-ng/package.py:80]: The problem in https://github.com/spack/spack/pull/44010#discussion_r1623078771 was that `zlib-ng+shared` wasn't actually building shared libraries, so this should add something, not just remove this fix:

* Maybe this package could raise an exception in `install()` if no `.so` is found
  * Or maybe we should do this universally for all packages with a `+shared` variant, since you observed it can happen more generally
* Maybe there should be a conflict for `+shared @:2.1.1` (or `@:2.1.6`, since it was pointed out above that this problem seems to affect more than `@:2.1.2`

Overall, I'm also not clear what you mean by "hide", because this change allowed the shared library to be built. Are you saying the library is not correct or usable? Regardless, I think this needs to do one of the above. 

If the library produced by the original fix is not broken then I think this PR should also be able to produce a shared library, so I'd still want this logic even if the error detection were added: I think it should be kept until we can clearly identify conditions under which we can get a functional `zlib-ng+shared@2.1.6:%cce` (i.e. adding constraints to the package.py) or provide a better way to get one (e.g. linker configuration/invocation)

**Review by scheibelp — DISMISSED** [review]

**Comment by haampie** [comment]
> Also, in my tests, Spack was using `zlib-ng@2.1.6`, so the issue that I saw cannot be the one fixed v2.1.2 (unless the issue was re-introduced in v2.1.6 after it was fixed in 2.1.2).

Can you reproduce that issue?

File: var/spack/repos/builtin/packages/zlib-ng/package.py (hunk 1 above)
Comment (haampie) [reply var/spack/repos/builtin/packages/zlib-ng/package.py:80]: If the build didn't fail after failure to create a shared library, then that's a build system problem. Report that upstream instead.

Comment (haampie) [reply var/spack/repos/builtin/packages/zlib-ng/package.py:80]: > not clear what you mean by "hide"

It is a legitimate problem for certain symbol versions to be referenced that do not exist. It could prevent LD_PRELOAD=/some/other/libz.so to work. You're disabling a strict check for that, so future regressions will not be caught.

Comment (scheibelp) [reply var/spack/repos/builtin/packages/zlib-ng/package.py:80]: > If the build didn't fail after failure to create a shared library, then that's a build system problem. 

I think it's our problem if there isn't an upstream fix (we often apply patches within the Spack repo that should be upstreamed, and we remove them when the upstream fix lands): we should avoid the broken versions (and make sure we can pick a combination that works for `zlib-ng+shared@2.1.6:%cce`), or we should detect failed installs, or we should keep this change.

> Report that upstream instead.

I don't think that has to be mutually exclusive with addressing things here in the meantime.

> It is a legitimate problem for certain symbol versions to be referenced that do not exist.

In other words:

1. If we keep everything as-is, we have a potentially broken `zlib-ng+shared` in some cases, but it exists (and seems usable), furthermore the constraint exists as an indicator of something wrong
2. If we apply everything in this PR as-is, then we have a broken `zlib-ng+shared`, and there is no indicator within the `zlib-ng` package that this might be the case

I think [1] is preferable for no other reason than there is at least some indication of something strange going on with `zlib-ng`. I think I've also suggested a couple reasonable alternatives to [1] and [2]

**Comment by v-dobrev** [comment]
> > Also, in my tests, Spack was using `zlib-ng@2.1.6`, so the issue that I saw cannot be the one fixed v2.1.2 (unless the issue was re-introduced in v2.1.6 after it was fixed in 2.1.2).
> 
> Can you reproduce that issue?

If you are referring to the issue I encountered, it is described here: https://github.com/spack/spack/pull/44010#discussion_r1628365884 -- it is from 2 weeks ago.

If you are referring to the issue that was resolved in `zlib-ng@2.1.2` in the PR that you linked above -- I have not tried that and I'm not sure how.

File: var/spack/repos/builtin/packages/zlib-ng/package.py (hunk 1 above)
Comment (haampie) [reply var/spack/repos/builtin/packages/zlib-ng/package.py:80]: > in some cases

you don't know which case

**Comment by haampie** [comment]
See https://github.com/zlib-ng/zlib-ng/pull/1750

**Review by alalazo — APPROVED** [review]
